Compute USB_DEVICES instead of requiring it to be specified

jepler opened the pull request:
If it is possible, it would be nice to avoid the need to change USB_DEVICES when changing CIRCUITPY_USB_HID and ... _MIDI.   If #3037 passes CI, then _this_ PR doesn't change anything about available USB devices, it just allows the value to be calculated in all cases.

jerryneedell commented:
I think this also has to include the atmel-samd pewpew10 build
see comment in #3037

jepler commented:
@tannewt Please take a look again, and at #3037 to see that the computed values and the hard-coded values agreed in all cases before I removed the hard-coded values and replaced them with the computed values.

jepler commented on supervisor/supervisor.mk:
```diff
@@ -115,12 +115,12 @@ ifndef USB_INTERFACE_NAME
-ifndef USB_HID_DEVICES
-USB_HID_DEVICES = "KEYBOARD,MOUSE,CONSUMER,GAMEPAD"
+ifeq ($(CIRCUITPY_USB_HID),1)
+USB_DEVICES := $(USB_DEVICES),HID
```
How about I change `+USB_DEVICES := "$(USB_DEVICES_COMPUTED)"` to `+USB_DEVICES ?= "$(USB_DEVICES_COMPUTED)"` so that it becomes a conditional assignment, and any value from a mpconfigboard.mk would override?

dhalbert replied:
If giving `USB_DEVICES` in `mpconfigboard.mk` would override it, would that also override the check? That's why I suggested adding `CIRCUITPY_USB_MIDI` and `CIRCUITPY_USB_HID` as toggles, so that the user never has to see `USB_DEVICES`.

btw, I don't understand the reason for using `:=` here instead of `=`. I know it makes an immediate instead of a deferred assignment, but would it not work with regular `=`?

Another check we could add (this is getting complicated) is counting endpoints for boards with limited endpoints, and warning you there are too many. It has to know CDC takes 2, and others take only one. It could count words/commas in `USB_DEVICES`.

jepler replied:
Yes, specifying USB_DEVICES manually would skip the check.

The `:=` is necessary when building `USB_DEVICES_CALCULATED` up in parts, because with the usual `=` syntax, this is a self-referential variable: `USB_DEVICES_COMPUTED := $(USB_DEVICES_COMPUTED),AUDIO`.  However, in the current version that only applies to USB_DEVICES_COMPUTED and not to USB_DEVICES, which is good, because there's no `?:=`.

Technically once you turn off MSC or CDC you don't have a CircuitPython compliant board anymore, so I don't know how far we want to go out of our way to enable this.  I can easily change the one spot to `?=` to enable it, but with the caveat that you also give up the automatic checking.  Can we leave other improvements to the future in order to get this improvement in place?

dhalbert replied:
> Can we leave other improvements to the future in order to get this improvement in place?

Yes, that's fine with me.
